### PR TITLE
Add an example to the readme?

### PR DESCRIPTION
**Pull request by mcabbott**
This proposes to add a very simple Flux model to the readme. Almost the simplest thing I could think of, which actually works. 

It's nice to see code right away. I guess it won't make sense if you've never seen a neural network, but if you've ever tried any demo elsewhere, this should look familiar, and point out the main features. Including (I hope) that batch dims are last, that the model is a mutable object, how to move to GPU, what arguments `train!` needs (and in what order). 

Without BatchNorm this works 90% of the time, with default Adam() too just 80%. So the post-1969 details are doing something.

On Julia 1.7+, just pasting this in will install everything including CUDA. ~~Maybe it should say that. I didn't want to put too many words.~~ Now it does.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,7 +2,13 @@
...
 </p>
 
-[![][action-img]][action-url] [![](https://img.shields.io/badge/docs-stable-blue.svg)](https://fluxml.github.io/Flux.jl/stable/) [![](https://img.shields.io/badge/chat-on%20slack-yellow.svg)](https://julialang.org/slack/) [![DOI](https://joss.theoj.org/papers/10.21105/joss.00602/status.svg)](https://doi.org/10.21105/joss.00602) [![ColPrac: Contributor's Guide on Collaborative Practices for Community Packages](https://img.shields.io/badge/ColPrac-Contributor's%20Guide-blueviolet)](https://github.com/SciML/ColPrac) [![][codecov-img]][codecov-url]
+<div align="center">
+
+[![](https://img.shields.io/badge/docs-stable-blue.svg)](https://fluxml.github.io/Flux.jl/stable/) [![](https://img.shields.io/badge/chat-on%20slack-yellow.svg)](https://julialang.org/slack/) [![ColPrac: Contributor's Guide on Collaborative Practices for Community Packages](https://img.shields.io/badge/ColPrac-Contributor's%20Guide-blueviolet)](https://github.com/SciML/ColPrac) [![DOI](https://joss.theoj.org/papers/10.21105/joss.00602/status.svg)](https://doi.org/10.21105/joss.00602)
+
+[![][action-img]][action-url] [![][codecov-img]][codecov-url]
```
Comment (mcabbott) [inline README.md:9]: Moved because these badges wrapped to 2 lines, and seemed to be in random order. This puts the CI ones on 2nd line, and centres them.

File: README.md [hunk 2]
```diff
@@ -12,7 +18,19 @@
...
-] add Flux
+using Flux
+
+x = hcat(digits.(0:3, base=2, pad=2)...) |> gpu
+y = Flux.onehotbatch(xor.(eachrow(x)...), 0:1) |> gpu
+data = ((Float32.(x), y) for _ in 1:100)
+
+model = Chain(Dense(2 => 3, sigmoid), BatchNorm(3), Dense(3 => 2), softmax) |> gpu
+optim = Adam(0.1, (0.7, 0.95))
+loss(x, y) = Flux.crossentropy(model(x), y)
```
Comment (mcabbott) [inline README.md:29]: Weirdly, using `logitcrossentropy` hurts performance:
```julia
julia> mean(1:100) do _
       model = Chain(Dense(2 => 3, sigmoid), BatchNorm(3), Dense(3 => 2), softmax) |> gpu
       optim = Adam(0.1, (0.7, 0.95))
       loss(x, y) = Flux.crossentropy(model(x), y)  # closes over model

       Flux.train!(loss, Flux.params(model), data, optim)

       all((model(x) .> 0.5) .== y)
       end
1.0

julia> mean(1:100) do _
       model2 = Chain(Dense(2 => 3, sigmoid), BatchNorm(3), Dense(3 => 2)) |> gpu
       optim2 = Adam(0.1, (0.7, 0.95))
       loss2(x, y) = Flux.logitcrossentropy(model2(x), y)  # this should be better right?

       Flux.train!(loss2, Flux.params(model2), data, optim2)

       all((model2(x) .> 0) .== y)
       end
0.81
```

Comment (ToucheSir) [reply README.md:30]: That seems concerning. Is this just initialization? Assuming some random seed is fixed before either of these are run, I wonder if flipping the order would make a difference.

Comment (mcabbott) [reply README.md:30]: These are fairly consistent, in any order, without ever resetting the seed. Same on 2 computers, and GPU.

Comment (mcabbott) [reply README.md:30]: Works the same with `Flux.crossentropy(model(x), y, ϵ=1f-1)` and `ϵ=0`. 

Works the same if I remove `@fastmath` and `if all(isfinite, max_)` from `logsoftmax!.

File: README.md (hunk 1 above)
Comment (Saransh-cpp) [reply README.md:9]: I think `FluxML` projects should also include the "number of downloads" badge in their READMEs -

[Flux Downloads](https://pkgs.genieframework.com?packages=Flux)
[Zygote Downloads](https://pkgs.genieframework.com?packages=Zygote)
[NNlib Downloads](https://pkgs.genieframework.com?packages=NNlib)
[Functors Downloads](https://pkgs.genieframework.com?packages=Functors)
[Metalhead Downloads](https://pkgs.genieframework.com?packages=Metalhead)
[FastAI Downloads](https://pkgs.genieframework.com?packages=FastAI)
... and so on

File: README.md [hunk 3]
```diff
@@ -2,7 +2,13 @@
...
 </p>
 
-[![][action-img]][action-url] [![](https://img.shields.io/badge/docs-stable-blue.svg)](https://fluxml.github.io/Flux.jl/stable/) [![](https://img.shields.io/badge/chat-on%20slack-yellow.svg)](https://julialang.org/slack/) [![DOI](https://joss.theoj.org/papers/10.21105/joss.00602/status.svg)](https://doi.org/10.21105/joss.00602) [![ColPrac: Contributor's Guide on Collaborative Practices for Community Packages](https://img.shields.io/badge/ColPrac-Contributor's%20Guide-blueviolet)](https://github.com/SciML/ColPrac) [![][codecov-img]][codecov-url]
+<div align="center">
+
+[![](https://img.shields.io/badge/docs-stable-blue.svg)](https://fluxml.github.io/Flux.jl/stable/) [![](https://img.shields.io/badge/chat-on%20slack-yellow.svg)](https://julialang.org/slack/) [![ColPrac: Contributor's Guide on Collaborative Practices for Community Packages](https://img.shields.io/badge/ColPrac-Contributor's%20Guide-blueviolet)](https://github.com/SciML/ColPrac) [![DOI](https://joss.theoj.org/papers/10.21105/joss.00602/status.svg)](https://doi.org/10.21105/joss.00602)
+<br/>
+[![][action-img]][action-url] [![][codecov-img]][codecov-url]
```
Comment (Saransh-cpp) [inline README.md:9]: Also, it might be better to arrange these badges in a table? Something like this -


| Type | Badge/Status |
| ----- | --------------- |
|   CI   | [![][action-img]][action-url] [![][codecov-img]][codecov-url] [pkgeval](https://juliahub.com/ui/Packages/Flux/QdkVy)|
| Documentation | [pages-build-deployment](https://github.com/FluxML/Flux.jl/actions/workflows/pages/pages-build-deployment) [](https://fluxml.github.io/Flux.jl/stable/) |
|Community| [](https://julialang.org/slack/) [Flux Downloads](https://pkgs.genieframework.com?packages=Flux) [ColPrac: Contributor's Guide on Collaborative Practices for Community Packages](https://github.com/SciML/ColPrac) [deps](https://juliahub.com/ui/Packages/Flux/QdkVy?t=2)|
|DOI and version| [DOI](https://doi.org/10.21105/joss.00602) [version](https://juliahub.com/ui/Packages/Flux/QdkVy) |

[action-img]: https://github.com/FluxML/Flux.jl/workflows/CI/badge.svg
[action-url]: https://github.com/FluxML/Flux.jl/actions
[codecov-img]: https://codecov.io/gh/FluxML/Flux.jl/branch/master/graph/badge.svg
[codecov-url]: https://codecov.io/gh/FluxML/Flux.jl

File: README.md (hunk 1 above)
Comment (mcabbott) [reply README.md:9]: Could do, 88k sounds like a lot...

File: README.md (hunk 3 above)
Comment (mcabbott) [reply README.md:9]: This is more logically organised. But I'm not sure these badges are really so valuable. Knowing to click the blue button is a quick way to get to the docs on an unknown package, but perhaps we should trim some of the others.

File: README.md (hunk 2 above)
Comment (mcabbott) [reply README.md:30]: Oh right, thanks! I'm an idiot...

**Comment by CarloLucibello** [comment]
The example feels too smart, but on the other hand it could be more catchy than a simpler mnist example.

**Comment by mcabbott** [comment]
I wondered about MNIST, but I think it'll end up a fair bit longer... and more of the code dealing with loading data etc, rather than core Flux + basic Julia stuff?

Could do linear regression pretty compactly, but that seems less illustrative.

I do think it should be short. Having a long nicely explained example means it starts competing with the docs as an alternative way to learn, and we already have too many paths. Just a minimal thing you can copy-paste on day 1 & see that everything is installed and working.

**Review by CarloLucibello — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -11,10 +17,23 @@
...
+
+x = hcat(digits.(0:3, base=2, pad=2)...) |> gpu  # let's solve the XOR problem!
+y = Flux.onehotbatch(xor.(eachrow(x)...), 0:1) |> gpu
+data = ((Float32.(x), y) for _ in 1:100)  # an iterator making Tuples
+
+model = Chain(Dense(2 => 3, sigmoid), BatchNorm(3), Dense(3 => 2)) |> gpu
+optim = Adam(0.1, (0.7, 0.95))
+mloss(x, y) = Flux.logitcrossentropy(model(x), y)  # closes over model
+
+Flux.train!(mloss, Flux.params(model), data, optim)  # updates model & optim
```
[the remark this answers is not in the document]
Comment (CarloLucibello) [reply README.md:32]: This seems the way forward to move `train!` away from `Params`.
